### ISSUE TITLE
Issues when upgrading to CanJS 4.0

### ISSUE THREAD

**Issue by BigAB** [canjs/can-stache#376]
### Prologue
So I just had a very confusing experience when trying to get some tests passing in `can-connect#major`.

### The failing test...
I had a test in `can-connect/can/tag` that used a template like this:

```
<person-model getList='{type=type}'>
	{{#isPending}}<span class='pending' on:inserted="pending()"></span>{{/isPending}}
	{{#isResolved}}<span class='resolved' on:inserted="resolved(%context, %element)">{{#each value}}<span>{{id}}</span>{{/each}}</span>{{/isResolved}}
</person-model>
```

which was giving me the following warning:
`WARN: can-stache-bindings: on:inserted couldn't find method named @pending {element: span.pending, scope: Scope}`
and 
`WARN: can-stache-bindings: on:inserted couldn't find method named @resolved {element: span.pending, scope: Scope}`

I combed through the test code to make sure it was right, I threw in some other properties on the viewModel/scope to see if they would print out, and they did, so I was baffled. I thought maybe it just needed to add some "explicit scope walking", but adding `../pending()` did't seem to help.

I was baffled.

### Getting help...
When I got help (from @bmomberger-bitovi) he inspected the scope object, and on the `_context` fields saw the tree, and then basically explained to me that the `{{#isPending}}` part in the template is actually creating a new "scope",  a scope thats `_context` is simply the primitive `true`. So I could fix it by either adding an additional `../` to the call expression or switch it to `{{#if isPending}}` which doesn't create a new scope.

Mind blown a little, I chose the latter, and we moved on.

### More trouble...
Then I got an error `Uncaught TypeError: Cannot read property 'childNodes' of undefined`, to which I found the offending line of code in the test:
```
equal(el.childNodes[1].innerHTML, "3", "added id");
```

Why wasn't element being passed to the call expression?

once again Brad pointed out that for the part`%scope` and `%element`, the `%` symbols are being dropped for the `scope` keyword. I was like, oh yeah I knew that, but I didn't think they would be removed outright. We changed then to `scope` and things started working again.

## The Point
Long story short _(too late)_, I am worried about the changes necessary to the templates in Can 4.0.  

I was aware of them and still struggled because I didn't know exactly what creates a new scope and what doesn't, and the warnings/errors didn't seem to tell me anything about what was wrong. From my point of view: my template, which previously worked, just stopped working when upgrading.

I was also surprised, maybe ignorantly, but surprised none-the-less that the `%` symbols were removed entirely in the next version. I had seen the warnings to change to `scope` recently, but didn't expect them to be removed so soon, so assumed something else was wrong.

So...
**I am concerned for the users who don't have as much access to stache experts as I do.**

Are these changes to harsh?
Is there anything we can do to help user through upgrading the stache templates for CanJS 4.0?

**Comment by justinbmeyer** [comment]
Everything here will have (or already does) have warnings in 3.X versions.  

The template will have worked, but have complained that those features will not work in 4.0.

Other than creating a migration guide (already planned), I don't think there's anything to do.  Users should upgrade to the latest 3.X version before migrating to 4.0.  Anyone who tries to go from something like 3.10 to 4.0 is making a mistake.

**Comment by justinbmeyer** [comment]
A philosophical note ... one of the big reasons to go 4.0 is so we can drop all these old ways of doing things.  While `scope.vars` and such are relatively recent, it's not worth keeping them around in 4.0 because:
- users would still see deprecation notices and 
- the OS team needs to support the additional complexity.

**Comment by bmomberger-bitovi** [comment]
> *I didn't know exactly what creates a new scope and what doesn't*

This is still a fundamental problem ill-addressed by Stache 4.0.  By removing scope walking, users are going to be frustrated by cases where scopes are created unexpectedly and their scope lookups do not work. There is no reliable indicator in Stache when a new scope is created: it can happen through a custom tag, a custom attribute, or a Stache section, but it is not mandatory in any of these cases.
